// javascript-binary-converter/build/cjs/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.utils = exports.HexStringConverter = exports.BinaryConverter = exports.NumberConverter = exports.DecimalBytesConverter = exports.BytesConverter = exports.ArrayBufferConverter = exports.ImageConverter = exports.FileConverter = exports.TypedArrayConverter = exports.BlobConverter = exports.converter = void 0;
const converter_1 = __importDefault(require("./converter"));
exports.converter = converter_1.default;
const BlobConverter_1 = __importDefault(require("./converters/BlobConverter"));
exports.BlobConverter = BlobConverter_1.default;
const TypedArrayConverter_1 = __importDefault(require("./converters/TypedArrayConverter"));
exports.TypedArrayConverter = TypedArrayConverter_1.default;
const FileConverter_1 = __importDefault(require("./converters/FileConverter"));
exports.FileConverter = FileConverter_1.default;
const ImageConverter_1 = __importDefault(require("./converters/ImageConverter"));
exports.ImageConverter = ImageConverter_1.default;
const ArrayBufferConverter_1 = __importDefault(require("./converters/ArrayBufferConverter"));
exports.ArrayBufferConverter = ArrayBufferConverter_1.default;
const BytesConverter_1 = __importDefault(require("./converters/BytesConverter"));
exports.BytesConverter = BytesConverter_1.default;
const DecimalBytesConverter_1 = __importDefault(require("./converters/DecimalBytesConverter"));
exports.DecimalBytesConverter = DecimalBytesConverter_1.default;
const NumberConverter_1 = __importDefault(require("./converters/NumberConverter"));
exports.NumberConverter = NumberConverter_1.default;
const BinaryConverter_1 = __importDefault(require("./converters/BinaryConverter"));
exports.BinaryConverter = BinaryConverter_1.default;
const HexStringConverter_1 = __importDefault(require("./converters/HexStringConverter"));
exports.HexStringConverter = HexStringConverter_1.default;
const utils = __importStar(require("./utils/index"));
exports.utils = utils;
exports.default = converter_1.default;
