// javascript-binary-converter/build/cjs/utils/string.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.stringToBinary = exports.splitStringToChunks = exports.isNumericString = exports.reverseString = exports.trimString = exports.padString = void 0;
function padString(string, requiredLength = 8, paddingCharacter = "0") {
    let fullString = string;
    const numOfZerosMissing = requiredLength - string.length;
    for (let i = 0; i < numOfZerosMissing; i++) {
        fullString = paddingCharacter + fullString;
    }
    return fullString;
}
exports.padString = padString;
function trimString(string, requiredLength = 8) {
    return string.slice(string.length - requiredLength);
}
exports.trimString = trimString;
function reverseString(s) {
    return s.split("").reverse().join("");
}
exports.reverseString = reverseString;
function isNumericString(n) {
    return !isNaN(parseFloat(n)) && isFinite(n) && typeof n === 'string';
}
exports.isNumericString = isNumericString;
function splitStringToChunks(string, chunkSize) {
    return string.match(new RegExp('.{1,' + chunkSize + '}', 'g'));
}
exports.splitStringToChunks = splitStringToChunks;
function stringToBinary(text) {
    const encoder = new TextEncoder();
    const typedArray = encoder.encode(text);
    let binary = "";
    for (let integer of typedArray) {
        binary += padString(integer.toString(2));
    }
    return binary;
}
exports.stringToBinary = stringToBinary;
